src/paginationCalls/pageDescriptor.ts

```diff
@@ -103,6 +116,108 @@ export class PageDescriptor implements Descriptor {
     return stream;
   }
 
+  /**
+   * Create an async iterable which can be recursively called for data on-demand.
+   */
+  asyncIterate(
+    apiCall: GaxCall,
+    request: RequestType,
+    options: CallSettings
+  ): AsyncIterable<{} | undefined> {
+    const iterable = this.createIterator(options);
+    const funcPromise =
+      typeof apiCall === 'function' ? Promise.resolve(apiCall) : apiCall;
+    funcPromise
+      .then((func: GaxCall) => {
+        this.resolveParams(request, func, options);
+      })
+      .catch(error => {
+        throw new Error(error);
+      });
+    return iterable;
+  }
+
+  createIterator(options: CallSettings): AsyncIterable<{} | undefined> {
+    const requestPageTokenFieldName = this.requestPageTokenField;
+    const responsePageTokenFieldName = this.responsePageTokenField;
+    const resourceField = this.resourceField;
+    const self = this;
+    const getNextPageRequest = this.getNextPageRequest;
+    const asyncIterable = {
+      [Symbol.asyncIterator]() {
+        const funcPromise = new Promise((resolve, reject) => {
+          self.resolveFunction = resolve;
+        });
+        const requestPromise = new Promise((resolve, reject) => {
+          self.resolveRequest = resolve;
+        });
+        const cache: Array<{}> = [];
+        let nextPageRequest: RequestType | null = {};
+        let firstCall = true;
+        return {
+          async next() {
+            const ongoingCall = new call.OngoingCallPromise(options.promise);
+            const func = (await funcPromise) as SimpleCallbackFunction;
+            const request = (await requestPromise) as RequestType;
+            if (cache.length > 0) {
+              const value = cache.shift();
+              return Promise.resolve({done: false, value});
+            }
+            if (!firstCall && !nextPageRequest) {
+              return Promise.resolve({done: true, value: -1});
+            }
+            nextPageRequest = await getNextPageRequest(
+              cache,
+              func,
+              firstCall ? request : nextPageRequest!,
+              ongoingCall,
+              requestPageTokenFieldName,
+              responsePageTokenFieldName,
+              resourceField
+            );
+            firstCall = false;
+            return Promise.resolve({done: false, value: cache.shift()});
```

Comment:
Is it possible that the next page request will return an empty page? In that case `cache.shift()` will be `undefined`.

Reply:
Logically if there's no more response, the previous `nextPageRequest` should be null for not sending request for the end of the pages. But in case, the return page is empty, added logic to check if the cache is empty after the elements have been added, if so, return `{done: true, value: -1}` if not, it implies that we still got some responses, return `{done: false, value: cache.shift()}` Thanks!